File: lc/DynamicProgramming/DP08LongestIncreasingSubsequence.java
package lc.DynamicProgramming;

public class DP08LongestIncreasingSubsequence {
    public static void main(String[] args) {
        DP08SolutionOne sol = new DP08SolutionOne();
        System.out.println("sol.lengthOfLIS(new int[]{7,7,7,7,7,7,7})"+sol.lengthOfLIS(new int[]{7,7,7,7,7,7,7}));
        assert sol.lengthOfLIS(new int[]{10,9,2,5,3,7,101,18}) == 4;
        assert sol.lengthOfLIS(new int[]{0,1,0,3,2,3}) == 4;
        assert sol.lengthOfLIS(new int[]{7,7,7,7,7,7,7}) == 1;
    }
}

class DP08SolutionOne {
    public int lengthOfLIS(int[] nums) {
        int[] lis = new int[nums.length];
        int result = 0;
        for(int i=0;i<nums.length;i++) {
            lis[i] = 1;
            for(int j=0;j<i;j++) {
                if(nums[j] < nums[i]) {
                    lis[i] = Math.max(lis[i], lis[j] + 1);
                }
            }
            result = Math.max(result, lis[i]);
        }
        return result;
    }
}

File: lc/DynamicProgramming/DP08LongestIncreasingSubsequence.mjs
// https://leetcode.com/problems/longest-increasing-subsequence/
// dynamic-programming, subsequence

import assert from "assert";

/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
    const n = nums.length;
    const dp = Array.from({ length: n }).fill(0);
    dp[0] = 1;
    for(let i = 1;i < n; i++) {
     dp[i] = 1;
     for(let j = i-1; j >= 0; j--)  {
        if(nums[j] < nums[i]) {
            dp[i] = Math.max(dp[i], 1 + dp[j]);
        }
     }
    }
    return Math.max(...dp);
};


assert(lengthOfLIS([10,9,2,5,3,7,101,18]) === 4);
assert(lengthOfLIS([0,1,0,3,2,3]) === 4);
assert(lengthOfLIS([7,7,7,7,7,7,7]) === 1);
